File: app/javascript/embeds/strava.js
import MediaSite from "media_site"

class Strava extends MediaSite {
  constructor() {
    super(
      /https:\/\/www\.strava\.com\/activities\/([0-9]+)(?:\/.*)?/,

      async (url) => {
        const match = url.match(this.regex)
        if (match) {
          const id = match[1]
          return id;
        }
        return null
      },

      (id) => {
        const wrapper = document.createElement("div")
        wrapper.style.display = "flex"
        wrapper.style.justifyContent = "center"

        const div = document.createElement("div")
        div.className = "strava-embed-placeholder"
        div.dataset.embedType = "activity"
        div.dataset.embedId = id
        div.dataset.style = "standard"
        div.dataset.fromEmbed = "false"

        const script = document.createElement('script')
        script.src = "https://strava-embeds.com/embed.js"

        wrapper.appendChild(div)
        wrapper.appendChild(script)
        return wrapper
      }
    )
  }
}

export default Strava

File: app/models/media_embed/providers/strava.rb
module MediaEmbed
  module Providers
    class Strava < Base
      REGEX = %r{\Ahttps://www\.strava\.com/activities/([0-9]+)}i

      def render(url)
        match = url.match(REGEX)
        return unless match

        placeholder = @view.tag.div(
          "",
          class: "strava-embed-placeholder",
          data: { embed_type: "activity", embed_id: match[1], style: "standard", from_embed: "false" }
        )
        script = @view.content_tag(:script, "", src: "https://strava-embeds.com/embed.js")

        @view.tag.div(@view.safe_join([ placeholder, script ]), style: "display: flex; justify-content: center")
      end
    end
  end
end
